helix-term/src/grammars.rs

```diff
@@ -0,0 +1,290 @@
+const REMOTE_NAME: &str = "helix-origin";
+
+pub fn fetch_grammars() {
+    run_parallel(get_grammar_configs(), fetch_grammar);
+}
+
+pub fn build_grammars() {
+    run_parallel(get_grammar_configs(), build_grammar);
+}
+
+fn run_parallel<F>(grammars: Vec<GrammarConfiguration>, job: F)
+where
+    F: Fn(GrammarConfiguration) + std::marker::Send + 'static + Copy,
+{
+    let mut n_jobs = 0;
+    let pool = threadpool::Builder::new().build();
+    let (tx, rx) = channel();
+
+    for grammar in grammars {
+        let tx = tx.clone();
+        n_jobs += 1;
+
+        pool.execute(move || {
+            job(grammar);
+
+            // report progress
+            tx.send(1).unwrap();
+        });
+    }
+    pool.join();
+
+    assert_eq!(rx.try_iter().sum::<usize>(), n_jobs);
+}
+
+pub fn fetch_grammar(grammar: GrammarConfiguration) {
+    if let GrammarSource::Git { remote, revision } = grammar.source {
+        let grammar_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
+            .join("../runtime/grammars/sources")
+            .join(grammar.grammar_id.clone());
+
+        fs::create_dir_all(grammar_dir.clone()).expect("Could not create grammar directory");
+
+        // create the grammar dir contains a git directory
+        if !grammar_dir.join(".git").is_dir() {
+            Command::new("git")
+                .args(["init"])
+                .current_dir(grammar_dir.clone())
+                .output()
+                .expect("Could not execute 'git'");
+        }
+
+        // ensure the remote matches the configured remote
+        if get_repository_info(&grammar_dir, vec!["remote", "get-url", REMOTE_NAME])
+            != Some(remote.clone())
+        {
+            set_remote(&grammar_dir, &remote);
+        }
+
+        // ensure the revision matches the configured revision
+        if get_repository_info(&grammar_dir, vec!["rev-parse", "HEAD"]) != Some(revision.clone()) {
+            // Fetch the exact revision from the remote.
+            // Supported by server-side git since v2.5.0 (July 2015),
+            // enabled by default on major git hosts.
+            Command::new("git")
+                .args(["fetch", REMOTE_NAME, &revision])
+                .current_dir(grammar_dir.clone())
+                .output()
+                .expect("Failed to execute 'git'");
+
+            Command::new("git")
+                .args(["checkout", &revision])
+                .current_dir(grammar_dir)
+                .output()
+                .expect("Failed to execute 'git'");
+
+            println!(
+                "Grammar '{}' checked out at '{}'.",
+                grammar.grammar_id, revision
+            );
+        } else {
+            println!("Grammar '{}' is already up to date.", grammar.grammar_id);
+        }
+    };
+}
+
+// Sets the remote for a repository to the given URL, creating the remote if
+// it does not yet exist.
+fn set_remote(repository: &Path, remote_url: &str) {
```

Comment:
This should be fallible. If this fails, it will just go ahead and probably fail when it moves onto the rev-parse, and it will be very confusing what the reason is.